src/algo/numeric.rs

```diff
@@ -231,6 +231,75 @@ pub fn exclusive_scan_inplace<Range, BinaryOp>(
     }
 }
 
+/// Computes the generalized inner product of two ranges using a combination and accumulation operation.
+///
+/// # Precondition
+///   - `[start1, end1)` represents valid positions in rng1.
+///   - `[start2, ...]` represents valid positions in rng2.
+///   - rng1 and rng2 have the same size for the specified range.
+///
+/// # Postcondition
+///   - Returns the generalized inner product of rng1 and rng2 using `combine_op` and `acc_op`.
+///   - Complexity: O(n) applications of `combine_op` and `acc_op`.
+///
+/// Where n == `rng1.distance(start1, end1)`.
+///
+/// # Example
+/// ```rust
+/// use stl::*;
+/// use rng::infix::*;
+///
+/// let rng1 = [1, 2, 3];
+/// let rng2 = [4, 5, 6];
+/// let start1 = rng1.start();
+/// let end1 = rng1.end();
+/// let start2 = rng2.start();
+/// let result = algo::inner_product(&rng1, start1, end1, &rng2, start2, 0, |x, y| x * y, |a, b| a + b);
+/// assert_eq!(result, 32); // (1*4) + (2*5) + (3*6) = 4 + 10 + 18 = 32
+/// ```
+///```rust
+/// use stl::*;
+/// use rng::infix::*;
+///
+/// let rng1 = [1, 2, 3];
+/// let rng2 = [4, 5, 6];
+/// let start1 = rng1.start();
+/// let end1 = rng1.end();
+/// let start2 = rng2.start();
+/// let result = algo::inner_product(&rng1, start1, end1, &rng2, start2, 0, |x, y| x * y, |a, b| a -b);
+/// assert_eq!(result, -32); // (1*4) - (2*5) -(3*6) = 0-4 - 10 - 18 = -32
+///```
+pub fn inner_product<Rng1, Rng2, T, CombineOp, AccOp>(
+    rng1: &Rng1,
+    mut start1: Rng1::Position,
+    end1: Rng1::Position,
+    rng2: &Rng2,
+    mut start2: Rng2::Position,
+    mut init: T,
+    combine_op: CombineOp,
+    acc_op: AccOp,
+) -> T
+where
+    Rng1: InputRange + ?Sized,
+    Rng2: InputRange<Element = Rng1::Element> + ?Sized,
+    T: Clone,
+    CombineOp: Fn(&Rng1::Element, &Rng2::Element) -> T,
```

Comment:
Introduce a new type param U, so that CombineOp: Fn(&Rng1::Element, &Rng2::Element) -> U